sql,schemachanger: disallow concurrent execution for new schema changes

thoszhang opened the pull request:
This PR prevents new-style schema changes from running concurrently with
any other schema changes, in two ways:

1. If there are mutations in progress (from either new or old schema
   changes) when attempting to plan a new-style schema change, we wait
   and poll until there are no mutations, and then restart the
   transaction.
2. If we try to write an old-style schema change job while there is a
   new-style schema change on the table, which is detected via a new
   field on the table descriptor for the new-style schema change job ID,
   an error is returned. This effectively prevents all schema changes,
   even the ones without mutations.

Most of this commit consists of testing. Testing knobs for the new
schema changer are introduced. We also now accumulate the statements
involved in the schema change, as strings, in the schema changer state
in `extraTxnState`. The executor now takes an argument to inject more
relevant state into the testing knobs, including the aforementioned
statements.

Release justification: Non-production code change (the new schema
changer is disabled for 21.1)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61042)
<!-- Reviewable:end -->

ajwerner commented:
Let's get @postamar reading these and starting to think about this.

postamar commented on pkg/sql/schemachanger/scjob/job.go:
```diff
@@ -162,6 +180,36 @@ func makeTargetStates(
+	txn *kv.Txn,
+	descriptors *descs.Collection,
+	descIDs []descpb.ID,
+	jobID jobspb.JobID,
```
This is almost the same function as `setDescriptorJobIDs`, perhaps they could both be unified as `CompareAndSetDescriptorJobIDs(..., expectedJobID jobspb.JobID, newJobID jobspb.JobID)`.

postamar reviewed (approved):
Looks good to me, just a few minor comments.

lucy-zhang commented on pkg/sql/schemachanger/scbuild/builder.go:
```diff
@@ -671,6 +680,49 @@ func (b *Builder) addNode(dir scpb.Target_Direction, elem scpb.Element) {
 }
 
+// getLockedTableDescriptor returns a table descriptor guaranteed to be "locked"
+// (i.e., with no other concurrent schema changes), or else a
```
The "locked" terminology comes from a (currently under-specified) concept of locking vs. non-locking changes to descriptors that @ajwerner and I have been discussing, where a locking change in a schema change corresponds roughly to a schema change that involves mutations and requires mutual exclusion. (Some long and not very digestible internal discussions: [1](https://cockroachlabs.slack.com/archives/CARM0PTK4/p1607110682076600), [2](https://cockroachlabs.slack.com/archives/DDQT4UW9K/p1614357598000200), [3](https://cockroachlabs.slack.com/archives/CARM0PTK4/p1613145290261200).)

I think the concept will need some work, but I'm inclined to keep the terminology for now. It is admittedly weird that the present implementation doesn't actually tracked "locked" descriptors in the builder at all, so it doesn't provide anything except possibly returning an error, because we assume all of our targets are for locking changes. I left some TODOs with some further thoughts on this.

I am open to changing the name of this method, though. Maybe `getTableDescriptorForLockingChange`?

postamar replied:
If the concept already exists then I'm cool with that. That wasn't obvious to me so I feel it would be worth defining that concept more explicitly in a comment somewhere, even if it's just:
```
// We define a _locking_ change in a schema change as corresponding roughly to a schema change that involves mutations and requires mutual exclusion.
```
I also like your idea of renaming the method like this because the current name implies that the descriptor is already locked.

lucy-zhang replied:
Renamed the method and added a comment.

ajwerner reviewed (commented):
nits and commentary from me, otherwise LGTM

> Reviewed 19 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-:-MUiULj15JNuaF7XwC9M:bxy7zf9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/sql/drop_table.go, line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiFiF6-URQhkOF9-4l:-MUiFiF6-URQhkOF9-4m:b-6cldem) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/drop_table.go#L432)):*
> ```Go
> 	if tableDesc.NewSchemaChangeJobID != 0 {
> 		return pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
> 			"cannot perform a schema change on table %q while it is undergoing a new-style schema change",
> ```

I wonder what ultimately we should do here. https://github.com/cockroachdb/cockroach/issues/61256 to track. 
___
*[pkg/sql/schema_change_plan_node.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiTz4A1eqQ69tEGSQV:-MUiTz4B6AOs9jI7hOdc:busmlv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L49)):*
> ```Go
> 	// undergoing other schema changes, wait for them to finish, and then restart
> 	// the transaction.
> 	if concurrentErr := (*scbuild.ConcurrentSchemaChangeError)(nil); errors.As(err, &concurrentErr) {
> ```

This handling is sort of subtle. Could you pull it out into a helper if only so that it can be a conduit for clearer commentary? 
Something like `maybeHanldeConcurrentSchemaChangeError` 

Can you also add a comment related to fairness. Namely that this waiting is not done in a fair way such that if there are multiple concurrent transactions operating over an overlapping set of descriptors but accessing them in different orders that this approach is likely to experience livelock. 
___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUiGZaq2zMbBdLgLdc4:b-qejlx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
> ```Go
> 			"schema change waiting for concurrent schema changes on descriptor %d",
> 			concurrentErr.DescriptorID())
> ```

We probably want to abort the transaction prior to waiting. Otherwise we may block other transactions. 

I think in terms of kv APIs, this is:
```
retryErr := p.txn.PrepareRetryableError(ctx, concurrentErr.Error())
txn.CleanupOnError(ctx, retryErr)
```
before the waiting. 

I do not think you need to call `ManualRestart`. I believe that the layer above should do that, but I could be wrong on that. 
___
*[pkg/sql/schemachanger/scbuild/builder.go, line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiQFrj1dBi8z4R_-mU:-MUiQFrk7hp_tuzxftZC:b2f7qn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schemachanger/scbuild/builder.go#L83)):*
> ```Go
> 	return buf.String()
> }
> ```

can you stick a `TODO` that it'd be better to accumulate a list of descriptors which have concurrent schema changes and that eventually we'd like to know about the largest set of descriptors we might be waiting for? We could potentially pull this set out at the end of building. 
___
*[pkg/sql/schemachanger/scexec/executor.go, line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiTcHr-qoqW-1u-vX2:-MUiTcHr-qoqW-1u-vX3:bq8afmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schemachanger/scexec/executor.go#L250)):*
> ```Go
> }
> 
> func UpdateDescriptorJobIDs(
> ```

comment 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUjEHUV7DLXxaZK5n2J:bmke1m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

We probably want to abort the transaction prior to waiting. Otherwise we may block other transactions. 

I think in terms of kv APIs, this is:
```
retryErr := p.txn.PrepareRetryableError(ctx, concurrentErr.Error())
txn.CleanupOnError(ctx, retryErr)
```
before the waiting. 

I do not think you need to call `ManualRestart`. I believe that the layer above should do that, but I could be wrong on that. 
</blockquote></details>

I was wrong. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUo2DJr0OEQUYyVlTk7:bjdf25c) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I was wrong. 
</blockquote></details>

Is there any hazard if you do both -- like create an error above the waiting and do `CleanupOnError` and then do the manual restart below?


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/sql/drop_table.go, line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiFiF6-URQhkOF9-4l:-MUibinf8dW5iRsIOpbR:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/drop_table.go#L432)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I wonder what ultimately we should do here. https://github.com/cockroachdb/cockroach/issues/61256 to track. 
</blockquote></details>

Thanks.
___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUo2e0o9OYB4zPHkSQ2:b-29h89c) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is there any hazard if you do both -- like create an error above the waiting and do `CleanupOnError` and then do the manual restart below?
</blockquote></details>

`CleanupOnError` finalizes the txn, so you can't call `ManualRestart` in that state. It results in a panic.
___
*[pkg/sql/schemachanger/scbuild/builder.go, line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiQFrj1dBi8z4R_-mU:-MUjKf04CVnCTsfbo2E1:b5aucqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schemachanger/scbuild/builder.go#L83)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you stick a `TODO` that it'd be better to accumulate a list of descriptors which have concurrent schema changes and that eventually we'd like to know about the largest set of descriptors we might be waiting for? We could potentially pull this set out at the end of building. 
</blockquote></details>

Done. I don't know if we'd want to actually "accumulate" them in the course of normal builder logic, though, since all that code assumes we have no mutations. Seems like if we want to do this, we should just figure out the set of descriptors that need locking updates at the outset, which is probably possible.
___
*[pkg/sql/schemachanger/scexec/executor.go, line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiTcHr-qoqW-1u-vX2:-MUjKaAZEfqYOp-etokk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schemachanger/scexec/executor.go#L250)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

comment 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
That also published my draft comments, so let me just push again.

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @lucy-zhang)

___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUo3UDw5B6OmCGmdt_l:b-t3shii) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

`CleanupOnError` finalizes the txn, so you can't call `ManualRestart` in that state. It results in a panic.
</blockquote></details>

gotcha @andreimatei any interest in getting in on this game?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @lucy-zhang)

___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUo3vjN-gkcWFWyBLLW:b-w4wbeg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

gotcha @andreimatei any interest in getting in on this game?
</blockquote></details>

The hazard here is that if we don't cleanup, then we'll be holding locks for an arbitrarily long period of time. 


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andreimatei)

___
*[pkg/sql/schema_change_plan_node.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiTz4A1eqQ69tEGSQV:-MUo74-5-mjJDJsi7Bt4:b-opgz87) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L49)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This handling is sort of subtle. Could you pull it out into a helper if only so that it can be a conduit for clearer commentary? 
Something like `maybeHanldeConcurrentSchemaChangeError` 

Can you also add a comment related to fairness. Namely that this waiting is not done in a fair way such that if there are multiple concurrent transactions operating over an overlapping set of descriptors but accessing them in different orders that this approach is likely to experience livelock. 
</blockquote></details>

I pulled this into a separate function.

We discussed the concurrent schema change scenario offline. This isn't a concern as such because the schema changes don't run into each other's "locks" before they commit.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andreimatei)

___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUsvTJn-QtsNFNM00ww:b6qp47i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The hazard here is that if we don't cleanup, then we'll be holding locks for an arbitrarily long period of time. 
</blockquote></details>

just spent a good while working through this with @andreimatei. I think we need to handle this rather far away from this code here. The stuff dealing with the two version invariant is quite bad. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andreimatei)

___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUsvtEx7ZGF1Lq0uSDB:b-uotpzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

just spent a good while working through this with @andreimatei. I think we need to handle this rather far away from this code here. The stuff dealing with the two version invariant is quite bad. 
</blockquote></details>

discussed offline:

So, the current thinking is that we should propagate this error up the call stack and then in `ex.makeErrEvent` we should treat this structured error as a retryable but shove it into the error payload. Then in `ex.txnStateTransitionApplyWrapper` detect the error in the error payload and wait for the appropriate descriptor -- then update the transaction accordingly. 


<!-- Sent from Reviewable.io -->

Commit:
sql,schemachanger: disallow concurrent execution for new schema changes

This PR prevents new-style schema changes from running concurrently with
any other schema changes, in two ways:

1. If there are mutations in progress (from either new or old schema
   changes) when attempting to plan a new-style schema change, we wait
   and poll until there are no mutations, and then restart the
   transaction.
2. If we try to write an old-style schema change job while there is a
   new-style schema change on the table, which is detected via a new
   field on the table descriptor for the new-style schema change job ID,
   an error is returned. This effectively prevents all schema changes,
   even the ones without mutations.

Most of this commit consists of testing. Testing knobs for the new
schema changer are introduced. We also now accumulate the statements
involved in the schema change, as strings, in the schema changer state
in `extraTxnState`. The executor now takes an argument to inject more
relevant state into the testing knobs, including the aforementioned
statements.

Release justification: Non-production code change (the new schema
changer is disabled for 21.1)

Release note: None

thoszhang reviewed (commented):
I got rid of the `AfterStage` testing knob since it's probably unsound for the executor to return an error if its ops have actually succeeded, but I didn't want to speculatively change its API either, given that it's unused.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-:-MUuzBkzBWTurVKIs6nD:b-969v3c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/schema_change_plan_node.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61042#-MUiGZapB-I4u45wy0-0:-MUuz3Ln7gpaDOdnDRfp:b-ji71fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d67cc556489ec400e10569028ee5745c88d5ce8a/pkg/sql/schema_change_plan_node.go#L53)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

discussed offline:

So, the current thinking is that we should propagate this error up the call stack and then in `ex.makeErrEvent` we should treat this structured error as a retryable but shove it into the error payload. Then in `ex.txnStateTransitionApplyWrapper` detect the error in the error payload and wait for the appropriate descriptor -- then update the transaction accordingly. 
</blockquote></details>

Done. I think this is more or less working.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
I think CI is running into https://github.com/cockroachdb/cockroach/pull/61471.

postamar commented:
This got merged as #64291.
